### PR TITLE
GitHub Actions Build: distcheck, check, verify-trees, coverage, coveralls

### PR DESCRIPTION
**Pull request by DaAwesomeP**
This is just a draft! Still lots to do, but it should go quickly once I fix this pesky GH Actions cancel issue.

Only problem right now is getting GitHub not to cancel the build after 45 minutes. It must be OOM or some other resource; GitHub usually allows jobs to run for a very, very long time.

Example cancelled build (not by me!):
 - Rerunning this one now, maybe it will work: https://github.com/DaAwesomeP/ola/actions/runs/4355751567
 - https://github.com/DaAwesomeP/ola/actions/runs/4339605290

### CODE REVIEW

File: .github/workflows/build.yaml [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+        if: runner.os == 'Linux'
+        # Silence all deprecated declarations on Linux due to auto_ptr making the build log too long
+        run: |
+          echo "GH_OLA_CONFIGURE_ARGS=$GH_OLA_CONFIGURE_ARGS CPPFLAGS=-Wno-deprecated-declarations" >> $GITHUB_ENV
+      - name: Print configure command
+        run: echo "./configure $GH_OLA_CONFIGURE_ARGS"
+      - name: Configure
+        run: ./configure $GH_OLA_CONFIGURE_ARGS
+      - name: distcheck
+        run: make distcheck -j${{ steps.num-cpu-cores.outputs.NUM_CPU_CORES }} VERBOSE=1
```
Comment (peternewman) [inline .github/workflows/build.yaml:64]: So have you tried this? Getting us just `make check` for now would be great if that is quick enough to not time out?
```suggestion
      - name: makecheck
        run: make check -j${{ steps.num-cpu-cores.outputs.NUM_CPU_CORES }} VERBOSE=1
```

There's also ccache which we've used successfully in the past and ought to speed things up if the cache is cached too.

Comment (DaAwesomeP) [reply .github/workflows/build.yaml:64]: I'll try simply `make check`!

What's weird is that it isn't timing out per say...it's almost like a watchdog has decided there hasn't been any output for a while and that is why it is stopping or some OOM situation. I think I need to open an issue elsewhere about it.

Yeah I'll look into ccache.

**Comment by DaAwesomeP** [comment]
OK, I tried `make check` and this is fascinating...it also gets stucking compiling somewhere around `libolausbdmx` and the `usbdmx` and `usbpro` plugins. Maybe I can try turning those off.

**Comment by DaAwesomeP** [comment]
OK, I tried and I am having an issue where the following command does not disable `usbdmx`:
```bash
./configure --enable-rdm-tests --enable-java-libs --enable-ja-rule --enable-e133 --disable-usbdmx --disable-usbpro CPPFLAGS=-Wno-deprecated-declarations
```
```
OLA Version 0.10.9

Prefix: '/usr/local'
Compiler: 'g++ -g -O2 -std=gnu++11 -pthread '
Linker: '/usr/bin/ld -m elf_x86_64   -ldl '
Python: /usr/bin/python3

Python API: yes
Java API: yes
Enable HTTP Server: yes
RDM Responder Tests: yes
Ja Rule: yes
Enabled Plugins: artnet dummy e131 espnet ftdidmx gpio karate kinet milinst opendmx openpixelcontrol osc pathport renard sandnet shownet spi stageprofi uartdmx usbdmx
UUCP Lock Directory: /var/lock
```
I am able to reproduce this locally.

**Comment by DaAwesomeP** [comment]
Hmmm, disabling these gets it further but not to completion. I'm thinking it must be running out of memory. I'll try disabling some of the parallelization.

**Comment by DaAwesomeP** [comment]
Alright, finally! It took 20 minutes to run, but disabling parallelization (forcing `-j1`) causes it to get to the end of `make check`. This might fix `distcheck` too!

One check does fail (`CredentialsTest)` because the tests are running as root. What is this test testing? Why can't it run as root?

**Comment by DaAwesomeP** [comment]
@peternewman @kripton Hey look at that! Fully working `make distcheck` and `make check`! I'm going to separate the verify trees into a separate job after `distcheck` and then this is good to go!

Looks like maybe nobody has run that verify trees check in a while.

I am no longer running `make dist` after `make distcheck` because it appears to be redundant: https://www.gnu.org/software/automake/manual/html_node/Preparing-Distributions.html

We still run `make check` separately because it also does gcov. It also has different configure flags that I copied from the Travis file.

**Comment by peternewman** [comment]
> **UPDATE:** It appears `ja-rule` causes this. Disabling for now.

Yeah sorry I could have told you that, it's because ja-rule needs the usbdmx plugin to talk to the interfaces and send DMX.

> Alright, finally! It took 20 minutes to run, but disabling parallelization (forcing `-j1`) causes it to get to the end of `make check`.

That's the same sort of runtime as Travis used to be.

> This might fix `distcheck` too!
🤞 

> One check does fail (`CredentialsTest)` because the tests are running as root. What is this test testing? Why can't it run as root?

Because running as root is bad and means our code is at risk of being responsible for a security breach. There's a configure option to run as root (which I assume skips that test), but not running the check as root would be better!

> @peternewman @kripton Hey look at that! Fully working `make distcheck` and `make check`! I'm going to separate the verify trees into a separate job after `distcheck` and then this is good to go!

Amazing, just the minimally working check/distcheck would be great progress for now! Indeed would probably be better for the short term than even verify trees etc and having them in a later PR.

> Looks like maybe nobody has run that verify trees check in a while.

Yeah our CI has been broken for quite a while! I missed some bits in the release because it was AWOL (and I didn't run it manually). Some of that stuff hasn't changed though, so I'm slightly suspicious about other results.

> I am no longer running `make dist` after `make distcheck` because it appears to be redundant: https://www.gnu.org/software/automake/manual/html_node/Preparing-Distributions.html

That seems about as I vaguely understood.

> We still run `make check` separately because it also does gcov. It also has different configure flags that I copied from the Travis file.

That makes sense, but again leaving the gcov until later would be fine and probably makes sense too.

Is it worth updating this issue as it sounds like you've found the cause? https://github.com/actions/runner-images/issues/7004

I wonder if our Docker usage has made it worse? Can we tell Docker to offer more memory/use more cores?

**Comment by DaAwesomeP** [comment]
> Because running as root is bad and means our code is at risk of being responsible for a security breach. There's a configure option to run as root (which I assume skips that test), but not running the check as root would be better!

Fixed, I now make a new user and run it the same way that the Debian build does.

> Amazing, just the minimally working check/distcheck would be great progress for now! Indeed would probably be better for the short term than even verify trees etc and having them in a later PR.

Separated it out already! It has the same results as before separating.

> Yeah our CI has been broken for quite a while! I missed some bits in the release because it was AWOL (and I didn't run it manually). Some of that stuff hasn't changed though, so I'm slightly suspicious about other results.

I think I'm not going to fix the source of these failures in this pull, especially because I (:sob:) need to rebase this to 0.10.

> That makes sense, but again leaving the gcov until later would be fine and probably makes sense too.

Nearly done with this too, just fixing the Coveralls upload bit. Should be done in an hour or so I think.

> Is it worth updating this issue as it sounds like you've found the cause? https://github.com/actions/runner-images/issues/7004

I know I should (and I will), though their lack of support with issues like these is not promising.

> I wonder if our Docker usage has made it worse? Can we tell Docker to offer more memory/use more cores?

Unsure. I can maybe try this out later.

**Comment by DaAwesomeP** [comment]
Testing this is a funny process. I have a chain of different commits trying different things every 5 minutes, but because it takes 20 to 30 minutes to run I have to keep track of what was what in many tabs. Before I solved the OOM issue I had to keep these tabs open or I would lose the logs too.

**Comment by DaAwesomeP** [comment]
OK, I think that I managed to re-target this to 0.10. This is good to go!

 - `distcheck` for both gcc and clang. This includes testing `make dist`, `make check`, `make install`, and `make clean`, and those artifacts are downloadable
 - `check` with gcov generation (`*.gcno` and `*.gcna` files), and those artifacts are downloadable
 - `coverage` for coverage reports with `gcovr`, and those reports are sent to Coveralls as well as nice HTML reports as artifacts are downloadable
 - Using `gcovr` means that coverage reports can be made offline and checked without needing Coveralls (or we can switch to other tools) in case they go the way of Travis (:-1:) or you want to check coverage before pushing
 - Verify trees to check `dist` output including a problem matcher for nice GitHub annotations
 - Not going to fix the verify trees issues in this pull; that is a larger feat and this already a large pull

@peternewman Please consider adding @coveralls as a collaborator to the repo and enabling the "LEAVE COMMENTS?" option so that it can report back coverage changes on pull requests (see https://github.com/coverallsapp/github-action/tree/release/v2#coveralls-comments-arent-added-to-my-pull-request).

File: .github/workflows/build.yaml [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+            default-jdk-headless maven
+      - name: Generate coverage
+        shell: bash
+        run: make coverage VERBOSE=1
+      - name: Upload coverage to Coveralls
+        uses: coverallsapp/github-action@v2
+        with:
+          # Coveralls GitHub action does not support its own format
+          # see: https://github.com/coverallsapp/github-action/issues/104
+          # file: coverage.coveralls.json
```
Comment (peternewman) [inline .github/workflows/build.yaml:151]:
```suggestion
          # file: coverage/coverage.coveralls.json
```

File: .github/workflows/build.yaml [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+          format: cobertura
+          flag-name: ${{ matrix.id }}
+      - name: Upload coverage artifacts
+        if: always()
+        uses: actions/upload-artifact@v3
+        with:
+          name: ola-${{ matrix.id }}-coverage
+          path: coverage/
+  verify-trees:
+    name: 'Verify trees for ${{ matrix.id }}'
```
Comment (peternewman) [inline .github/workflows/build.yaml:162]: I'd rather this didn't run while its returning what I suspect are false positives. Can we either comment it out or removed the commits to a future PR (but don't bother with the latter unless it's trivial).

I think the issue is either this build being more complete, or us not properly cleaning up the contents of nodist, but I'm not sure which is technically the problem.

File: configure.ac [hunk 4]
```diff
@@ -1036,4 +1040,5 @@ Now type 'make @<:@<target>@:>@'
     check        - run the tests
     doxygen-doc  - generate the HTML documentation
     lint         - run the linters
+    coverage     - generate coverage post --enable-gcov
```
Comment (peternewman) [inline configure.ac:1043]: I think this should be able to force --enable-gcov like we do with say ja-rule and the USB stuff.

**Review by peternewman — CHANGES_REQUESTED** [review]
I'd be quite happy to just get one or more of check or distcheck in first, and review the other stuff in due course.

**Comment by peternewman** [comment]
> @peternewman Please consider adding @coveralls as a collaborator to the repo and enabling the "LEAVE COMMENTS?" option so that it can report back coverage changes on pull requests (see https://github.com/coverallsapp/github-action/tree/release/v2#coveralls-comments-arent-added-to-my-pull-request).

I think I turned off the coveralls comments last time as they were a bit spammy. You can still see the stats in the checks and we could presumably mandate a minimum level to pass...

File: .github/workflows/build.yaml (hunk 3 above)
Comment (DaAwesomeP) [reply .github/workflows/build.yaml:162]: Removing the commits will not be trivial; I'll see if I can mark as OK-to-fail or I'll comment it out. We did a similar thing for spellintian (still runs but issues only warnings and returns success).

File: configure.ac (hunk 4 above)
Comment (DaAwesomeP) [reply configure.ac:1043]: `--enable-gcov` is a `configure` flag, but running `make coverage` would happen after the fact. Are you saying that running `make coverage` should re-run `configure` with `--enable-gcov` added?

File: .github/workflows/build.yaml [hunk 5]
```diff
@@ -0,0 +1,204 @@
...
+    runs-on: ubuntu-20.04
+    container: ${{ matrix.container }}
+    timeout-minutes: 360
+    strategy:
+      fail-fast: false
+      matrix:
+        include:
+          - { id: "coverage-debian-stable-amd64-gcc",    task: "check",     configure-args: "--enable-ja-rule --enable-e133 --enable-gcov",                         container: "debian:stable", compiler: { CC: "/usr/bin/gcc",   CXX: "/usr/bin/g++",     pkg: "gcc g++" } }
+          - { id: "distcheck-debian-stable-amd64-gcc",   task: "distcheck", configure-args: "--enable-ja-rule --enable-e133 --enable-rdm-tests --enable-java-libs", container: "debian:stable", compiler: { CC: "/usr/bin/gcc",   CXX: "/usr/bin/g++",     pkg: "gcc g++" } }
+          - { id: "distcheck-debian-stable-amd64-clang", task: "distcheck", configure-args: "--enable-ja-rule --enable-e133 --enable-rdm-tests --enable-java-libs", container: "debian:stable", compiler: { CC: "/usr/bin/clang", CXX: "/usr/bin/clang++", pkg: "clang"   } }
```
[the remark this answers is not in the document]
Comment (DaAwesomeP) [reply .github/workflows/build.yaml]: Fixed.

File: .github/workflows/build.yaml (hunk 3 above)
Comment (DaAwesomeP) [reply .github/workflows/build.yaml:183]: Fixed.

Comment (DaAwesomeP) [reply .github/workflows/build.yaml:183]: Build succeeds now!

Comment (peternewman) [reply .github/workflows/build.yaml:183]: I quite liked the spellintian solution, or leaving the matcher but matching as info, but this is also fine.

File: .github/workflows/build.yaml (hunk 5 above)
[the remark this answers is not in the document]
Comment (peternewman) [reply .github/workflows/build.yaml]: I hadn't really clocked that, and it does odd wrapping for me on the GitHub diff (e.g. the space after CC). I think you can still see quite clearly e.g. which configure lines differ without too much trouble.

File: .github/workflows/build.yaml [hunk 6]
```diff
@@ -0,0 +1,204 @@
...
+        run: |
+          tar -xvzf ola-${{ matrix.id }}-source-tree.tar.gz .
+          rm ola-${{ matrix.id }}-source-tree.tar.gz
+      - name: Display structure of extracted files
+        if: env.ACTIONS_STEP_DEBUG == 'true'
+        run: ls -alR
+      - name: Update package database
+        run: apt-get update -y
+      - name: Install coverage tools
+        run: apt-get -y install curl gcc python3-pip
```
[the remark this answers is not in the document]
Comment (peternewman) [reply .github/workflows/build.yaml:153]: Cool, I was just checking we shouldn't be using the compiler arg from the earlier matrix config.

File: configure.ac (hunk 4 above)
Comment (peternewman) [reply configure.ac:1043]: Ah gotcha. Yeah maybe not that, but presumably we could at least throw a different error if `--enable-gcov` wasn't passed to configure and we call `make coverage`, given it presumably won't work in this case.

Comment (DaAwesomeP) [reply configure.ac:1043]: Ah yes, we can certainly catch that. I am honestly not very familiar with automake so I'm not certain how to do that.

Comment (DaAwesomeP) [reply configure.ac:1043]: OK, I think I got this to work. Running `make coverage` should now build if it needs to and then run the coverage generation (similar to how `make check` builds and runs unit tests).

Comment (DaAwesomeP) [reply configure.ac]: OK, this is passing! Running `make check` runs the checks and then the coverage. I couldn't figure out how to get it to compile the checks without running them, but realistically this isn't very necessary.

**Comment by DaAwesomeP** [comment]
Hey @peternewman anything else to change or are we good to go?

**Review by peternewman — COMMENTED** [review]
Sorry a few more comments...

File: .github/workflows/build.yaml [hunk 7]
```diff
@@ -0,0 +1,188 @@
...
+        # Silence all deprecated declarations on Linux due to auto_ptr making the build log too long
+        run: |
+          echo "GH_OLA_CONFIGURE_ARGS=$GH_OLA_CONFIGURE_ARGS CPPFLAGS=-Wno-deprecated-declarations" >> $GITHUB_ENV
+      - name: Print configure command
+        run: echo "./configure $GH_OLA_CONFIGURE_ARGS"
+      - name: Configure
+        run: sudo -u builduser ./configure $GH_OLA_CONFIGURE_ARGS
+      - name: ${{ matrix.task }}
+        # TODO(Perry): Parallelization causes GH Actions to hang -j${{ steps.num-cpu-cores.outputs.NUM_CPU_CORES }}
+        run: sudo -u builduser make ${{ matrix.task }} -j1 VERBOSE=1
```
[the remark this answers is not in the document]
Comment (DaAwesomeP) [reply .github/workflows/build.yaml]: Fixed!

File: .github/workflows/build.yaml [hunk 8]
```diff
@@ -0,0 +1,188 @@
...
+      - name: Find dist build tarball
+        run: |
+          echo "GH_OLA_VERIFY_TREES_TARBALL=$(ls -t --time=birth ola*.tar.gz| head -1)" >> $GITHUB_ENV
+      - name: Print dist build tarball name
+        run: echo "$GH_OLA_VERIFY_TREES_TARBALL"
+      - name: Extract dist build
+        run: tar -xvzf $GH_OLA_VERIFY_TREES_TARBALL
+      - name: Verify trees
+        shell: bash
+        # Always succeed for now until verify trees is fixed
```
[the remark this answers is not in the document]
Comment (DaAwesomeP) [reply .github/workflows/build.yaml]: Fixed!

**Comment by DaAwesomeP** [comment]
Debian 12 is now Debian stable, and it handles Python 3 system dependencies/pip differently, so I need to fix that. Working on that now!

**Comment by DaAwesomeP** [comment]
OK, I think I fixed the Python3 venv/system-wide issue (all builds are passing the `configure` stage). We'll know for sure in about 25 minutes.

Here is what needed to change:

1. Debian 12 no longer allows pip to install packages system-wide alongside `python3-*` packages installed by APT.
2. We now create a clean `venv` for the build. This `venv` is set to also use system packages, so we can still use the Debian-provided packages as they would be used for building. Like before, we also install the newer tools that we want from pip.
3. When we run the build as `builduser` (since we can't run as `root` or we fail an OLA test), we must pass along the `$PATH` variable (which activates the `venv`) and also the `CC`/`CXX` variables. The `--preserve-env` flag won't pass `$PATH`, so we pass that with the `env` helper command.

**Comment by DaAwesomeP** [comment]
Passing! I was simply missing `id:` on the number of cores step.

File: .github/workflows/build.yaml [hunk 9]
```diff
@@ -0,0 +1,199 @@
...
+          tar -xvzf ola-${{ matrix.id }}-source-tree.tar.gz .
+          rm ola-${{ matrix.id }}-source-tree.tar.gz
+      - name: Display structure of extracted files
+        if: env.ACTIONS_STEP_DEBUG == 'true'
+        run: ls -alR
+      - name: Update package database
+        run: apt-get update -y
+      - name: Install Python
+        run: apt-get -y install python3 python-is-python3
+      # Disable problem matcher for now until verify trees is fixed
```
Comment (peternewman) [inline .github/workflows/build.yaml:185]:
```suggestion
      # TODO(Perry): Disable problem matcher for now until verify trees is fixed
```

**Review by peternewman — APPROVED** [review]
Just the one TODO comment which I'll auto merge.

Thanks again for this @DaAwesomeP ! It's awesome!

File: .github/workflows/build.yaml (hunk 9 above)
Comment (DaAwesomeP) [reply .github/workflows/build.yaml]: Committed!

Comment (peternewman) [reply .github/workflows/build.yaml]: Thanks @DaAwesomeP . It wouldn't let me edit your code, maybe you've disabled contributor access?

**Comment by peternewman** [comment]
Coverage wasn't in the required list, but I figure it probably doesn't need to be unless we're being really strict!

I guess we'll now see if it pushes our coverage status correctly...

**Comment by DaAwesomeP** [comment]
> Coverage wasn't in the required list, but I figure it probably doesn't need to be unless we're being really strict!

It might be possible to setup rules that block merging is coverage dips by X percent in total? I'm not sure.

> I guess we'll now see if it pushes our coverage status correctly...

Passed! See https://coveralls.io/github/OpenLightingProject/ola?branch=0.10 and https://coveralls.io/jobs/123479842

**Comment by peternewman** [comment]
> > Coverage wasn't in the required list, but I figure it probably doesn't need to be unless we're being really strict!
> 
> It might be possible to setup rules that block merging is coverage dips by X percent in total? I'm not sure.

Maybe if we can just get it to comment if it drops by say 5% or something? Then it will only be usefully spammy!

**Pull request merged**